lib/linked_list.rb

```diff
@@ -12,150 +12,266 @@ def initialize(value, next_node = nil)
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def reverse
+    prev = nil
+    next_node = nil
+    current = @head
+    while current != nil
+     
+      next_node = current.next
+      current.next = prev
+      prev = current
+      current = next_node
+    end
+    @head = prev
+  end
 
-    ## Advanced Exercises
-    # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def find_middle_value
-      raise NotImplementedError
+  ## Advanced Exercises
+  # returns the value at the middle element in the singly linked list
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def find_middle_value
+    return nil if !@head
+    slow = @head
+    fast = @head
+    while fast.next != nil
+      slow = slow.next
+      fast = fast.next.next
     end
+    return slow.data
+  end
 
-    # find the nth node from the end and return its value
-    # assume indexing starts at 0 while counting to n
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def find_nth_from_end(n)
-      raise NotImplementedError
+  # find the nth node from the end and return its value
+  # assume indexing starts at 0 while counting to n
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def find_nth_from_end(n)
+    return nil if !@head
+    current = @head
+    index = 1
+    while current != nil
+      if length - index == n
+        return current.data
+      end
+      index += 1
+      current = current.next
     end
+    return nil
+  end
 
-    # checks if the linked list has a cycle. A cycle exists if any node in the
-    # linked list links to a node already visited.
-    # returns true if a cycle is found, false otherwise.
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def has_cycle
-      raise NotImplementedError
+  # checks if the linked list has a cycle. A cycle exists if any node in the
+  # linked list links to a node already visited.
+  # returns true if a cycle is found, false otherwise.
+  # Time Complexity: ?
+  # Space Complexity: ?
+  def has_cycle
+    return false if !@head || @head.next = nil
+    slow = @head
+    fast = @head
+    while fast != nil
+      slow = slow.next
+      fast = fast.next
+      if fast != nil
+        fast = fast.next
+      end
+      return true if fast == slow
     end
+    return false
+  end
 
+  # Additional Exercises
+  # returns the value in the first node
+  # returns nil if the list is empty
+  # Time Complexity: O(1)
+  # Space Complexity: O(1)
+  def get_first
+    return @head.data if @head
+    return nil
+  end
 
-    # Additional Exercises 
-    # returns the value in the first node
-    # returns nil if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def get_first
-      raise NotImplementedError
+  # method that inserts a given value as a new last node in the linked list
+  # Time Complexity: O(1)
+  # Space Complexity: O(1)
+  def add_last(value)

```

Comment:
This method works, but since you traverse the list until the end of the list... the time complexity is O(n)